<!doctype html><html>
<head>
<meta charset=utf-8>
<meta http-equiv=x-ua-compatible content="IE=edge"><title>确定TopHat中mate-inner-dist参数 - Yulong Niu</title><link rel=icon type=image/png href=icons/favicon.ico><meta name=viewport content="width=device-width,initial-scale=1">
<meta property="og:title" content="确定TopHat中mate-inner-dist参数">
<meta property="og:description" content="跳转至HISAT2或kallisto。">
<meta property="og:type" content="article">
<meta property="og:url" content="https://YulongNiu.github.io/posts/2015-05-23-tophat-inner-distance/"><meta property="article:section" content="posts">
<meta property="article:published_time" content="2015-05-23T19:08:19+08:00">
<meta property="article:modified_time" content="2015-05-23T19:08:19+08:00">
<meta name=twitter:card content="summary">
<meta name=twitter:title content="确定TopHat中mate-inner-dist参数">
<meta name=twitter:description content="跳转至HISAT2或kallisto。">
<link href="https://fonts.googleapis.com/css?family=Ubuntu:300,400,300italic,400italic|Raleway:200,300" rel=stylesheet>
<link rel=stylesheet type=text/css media=screen href=https://YulongNiu.github.io/css/bootstrap-table.css>
<link rel=stylesheet type=text/css media=screen href=https://YulongNiu.github.io/css/normalize.css>
<link rel=stylesheet type=text/css media=screen href=https://YulongNiu.github.io/css/main.css>
<link rel=stylesheet type=text/css href=https://YulongNiu.github.io/css/custom.css>
<link rel=stylesheet type=text/css href=https://YulongNiu.github.io/css/dark.css media="(prefers-color-scheme: dark)">
<link rel=stylesheet type=text/css href=https://YulongNiu.github.io/css/custom-dark.css media="(prefers-color-scheme: dark)">
<script src=https://cdn.jsdelivr.net/npm/feather-icons/dist/feather.min.js></script>
<script src=https://YulongNiu.github.io/js/main.js></script>
<script src=https://YulongNiu.github.io/js/abc.js></script>
<script src=https://YulongNiu.github.io/js/xyz.js></script>
</head>
<body>
<div class="container wrapper post">
<div class=header>
<base href=https://YulongNiu.github.io/>
<h1 class=site-title><a href=https://YulongNiu.github.io/>Yulong Niu</a></h1>
<div class=site-description><h2>个人博客</h2><nav class="nav social">
<ul class=flat><a href=https://github.com/YulongNiu title=Github><i data-feather=github></i></a><a href=https://twitter.com/YulongNiu title=Twitter><i data-feather=twitter></i></a></ul>
</nav>
</div>
<nav class=nav>
<ul class=flat>
<li>
<a href=/>Home</a>
</li>
<li>
<a href=/posts>Posts</a>
</li>
<li>
<a href=/tags>Tags</a>
</li>
<li>
<a href=/about>About</a>
</li>
</ul>
</nav>
</div>
<div class=post-header>
<h1 class=title>确定TopHat中mate-inner-dist参数</h1>
<div class=meta>Posted at &mdash; May 23, 2015</div>
</div>
<div class=markdown>
<p>对于双端测序RNA-seq数据，TopHat在运行时候，有两个参数<code>-r/--mate-inner-dist</code>和<code>--mate-std-dev</code>分别标识一对reads的间隔长度的期望平均值和标准差，其默认值分别为50bp和20bp。这两个参数本身是个估计值，用于TopHat在map过程中确定一对reads是否匹配到基因组正确位置。如果能够准确设定这两个数值，将会提升TopHat结果的准确性和完整性，参考<a href=http://yulongniu.bionutshell.org/blog/2014/07/23/rna-seq-analysis/>一个例子</a>。</p>
<p>有两种方法获得这对参数的准确值：</p>
<p>第一种：获取RNA-seq实验建库方法，之后按照以下网址说明计算，<a href=http://blog.qiuworld.com:8080/archives/3007>RNA-seq差异表达分析工作流程</a>。</p>
<p>第二种：根据RNA-seq数据进行估算，具体步骤为：</p>
<ol>
<li>
<p>使用TopHat默认参数先跑一遍。</p>
</li>
<li>
<p>使用<a href=http://miso.readthedocs.org/en/fastmiso/#computing-the-insert-length-distribution-and-its-statistics>MISO</a>的<code>pe_utils</code>工具估算。</p>
</li>
</ol>
<p>以下详细介绍<code>pe_utils</code>使用方法。</p>
<p>第一步， 下载对应物种的基因注释文件GTF或者GFF，比如<a href="http://genome.ucsc.edu/cgi-bin/hgTables?command=start">USCS Table Browser</a>（<code>output format</code>选择<code>GTF</code>）或者使用MISO提供的<a href=http://miso.readthedocs.org/en/fastmiso/#human-mouse-gene-models-for-isoform-centric-analyses>Ensembl版本</a>。如果GTF文件，使用<a href=https://cole-trapnell-lab.github.io/cufflinks/file_formats/>Cufflinks</a>的<code>gffread</code>工具进行转换。</p>
<p>第二步，确定TopHat运行结果的bam文件与基因注释GFF文件，两者基因组命名方法一致。有的使用类似<code>chr1</code>命名，而另外一些使用<code>1</code>。如果不一致，建议修改GFF文件。</p>
<div class=highlight><pre tabindex=0 style=background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=color:green># 查看GFF文件中基因组命名</span>
$ awk <span style=color:#a31515>&#39;{print $1}&#39;</span> hg19USCS.gff | sort -n | uniq -c

<span style=color:green># 查看bam文件中基因组命名</span>
samtools view accepted_hits.bam | head -1000 | awk <span style=color:#a31515>&#39;{print $3}&#39;</span> | sort -n | uniq -c
</code></pre></div><p>第三步，筛选较长外显子，比如长度大于1000bp。MISO提供了<code>exon_utils</code>工具用于提取长外显子，但是我们没有能够成功运行过。因此这里提供一个R版本的脚本，比如基因注释文件名为<code>hg19USCS.gff</code>，输出筛选的文件名为<code>hg19USCS_selected.gff</code>。</p>
<div class=highlight><pre tabindex=0 style=background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash>gffFile &lt;- read.table(<span style=color:#a31515>&#39;hg19USCS.gff&#39;</span>, stringsAsFactors = FALSE)

gffExon &lt;- gffFile[gffFile[, 3] == <span style=color:#a31515>&#39;exon&#39;</span>, ]

exonLen &lt;- abs(gffExon[, 5] - gffExon[, 4])

gffExonSelect &lt;- gffExon[exonLen &gt;= 1000, ]

write.table(gffExonSelect, <span style=color:#a31515>&#39;hg19USCS_selected.gff&#39;</span>, 
            row.names = FALSE, col.names = FALSE,
            quote = FALSE, sep = <span style=color:#a31515>&#39;\t&#39;</span>)
</code></pre></div><p>第四步，使用<code>pe_utils</code>，实例如下：</p>
<div class=highlight><pre tabindex=0 style=background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=color:green>## 输入bam文件和GFF文件</span>
$ pe_utils --compute-insert-len accepted_hits.bam hg19USCS_selected.gff --output-dir insert-dist/

<span style=color:green>## 在insert-dist会出现类似accepted_filtered.bam.insert_len文件</span>
<span style=color:green>## -r/--mate-inner-dist估计值为mean</span>
<span style=color:green>## --mate-std-dev估计值为sdev</span>
$ head -1 accepted_filtered.bam.insert_len
<span style=color:green>## mean=165.3,sdev=45.2,dispersion=3.5,num_pairs=5622239</span>
</code></pre></div><h3 id=更新记录>更新记录</h3>
<p>2015年5月23日</p>
</div>
<div class=post-tags>
<nav class="nav tags">
<ul class=flat>
<li><a href=/tags/bioinfo>bioinfo</a></li>
</ul>
</nav>
</div></div>
<div class="footer wrapper">
<nav class=nav>
<div> © Copyright notice | <a href=https://github.com/vividvilla/ezhil>Ezhil theme</a> | Built with <a href=https://gohugo.io>Hugo</a></div>
</nav>
</div>
<script type=application/javascript>var doNotTrack=!1;doNotTrack||(window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga('create','UA-53073828-1','auto'),ga('send','pageview'))</script>
<script async src=https://www.google-analytics.com/analytics.js></script>
<script>feather.replace()</script>
</body>
</html>