<!DOCTYPE html>
<html>
<head>
	<meta charset="utf-8" />
	<meta http-equiv="X-UA-Compatible" content="IE=edge"><title>估计RNA-seq转录本表达量和寻找差异表达基因 - Yulong Niu</title><link rel="icon" type="image/png" href=icons/favicon.ico /><meta name="viewport" content="width=device-width, initial-scale=1">
	<meta property="og:title" content="估计RNA-seq转录本表达量和寻找差异表达基因" />
<meta property="og:description" content="RNA转录本计数。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://YulongNiu.github.io/posts/2015-06-03-quantifing-rna-seq/" />
<meta property="article:published_time" content="2015-06-03T16:58:01+08:00" />
<meta property="article:modified_time" content="2015-06-03T16:58:01+08:00" />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="估计RNA-seq转录本表达量和寻找差异表达基因"/>
<meta name="twitter:description" content="RNA转录本计数。"/>
<link href="https://fonts.googleapis.com/css?family=Ubuntu:300,400,300italic,400italic|Raleway:200,300" rel="stylesheet">
  <link rel="stylesheet" type="text/css" media="screen" href="https://YulongNiu.github.io/css/bootstrap-table.css" />
	<link rel="stylesheet" type="text/css" media="screen" href="https://YulongNiu.github.io/css/normalize.css" />
	<link rel="stylesheet" type="text/css" media="screen" href="https://YulongNiu.github.io/css/main.css" />
	<link rel="stylesheet" type="text/css" href="https://YulongNiu.github.io/css/custom.css" />
	<link rel="stylesheet" type="text/css" href="https://YulongNiu.github.io/css/dark.css" media="(prefers-color-scheme: dark)" />
	<link rel="stylesheet" type="text/css" href="https://YulongNiu.github.io/css/custom-dark.css" media="(prefers-color-scheme: dark)" />

	
	
	<script src="https://cdn.jsdelivr.net/npm/feather-icons/dist/feather.min.js"></script>
	<script src="https://YulongNiu.github.io/js/main.js"></script>
	<script src="https://YulongNiu.github.io/js/abc.js"></script>
	<script src="https://YulongNiu.github.io/js/xyz.js"></script>
</head>

<body>
	<div class="container wrapper post">
		<div class="header">
    <base href="https://YulongNiu.github.io/">
	<h1 class="site-title"><a href="https://YulongNiu.github.io/">Yulong Niu</a></h1>
	<div class="site-description"><h2>个人博客</h2><nav class="nav social">
			<ul class="flat"><a href="https://github.com/YulongNiu" title="Github"><i data-feather="github"></i></a><a href="https://twitter.com/YulongNiu" title="Twitter"><i data-feather="twitter"></i></a></ul>
		</nav>
	</div>

	<nav class="nav">
		<ul class="flat">
			
			<li>
				<a href="/">Home</a>
			</li>
			
			<li>
				<a href="/posts">Posts</a>
			</li>
			
			<li>
				<a href="/tags">Tags</a>
			</li>
			
			<li>
				<a href="/about">About</a>
			</li>
			
		</ul>
	</nav>
</div>


		<div class="post-header">
			<h1 class="title">估计RNA-seq转录本表达量和寻找差异表达基因</h1>
			<div class="meta">Posted at &mdash; Jun 3, 2015</div>
		</div>

		<div class="markdown">
			<h2 id="1-标准化和计数">1. 标准化和计数</h2>
<h3 id="11-htseq">1.1 HTseq</h3>
<p><strong>简介</strong>：<a href="http://www-huber.embl.de/users/anders/HTSeq/doc/index.html">HTseq</a> 是用于Python平台写成的处理高通量测序的平台。<code>htseq-count</code>可以用来对原始转录本计数，具体计数规则参考<a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">Counting reads in features with htseq-count</a>。</p>
<p><strong>平台</strong>：Python跨平台使用。</p>
<p><strong>快速运行</strong>：</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash"><span style="color:#008000"># 查看帮助</span>
$ htseq-count --help

<span style="color:#008000"># 对bam文件按照reads名称排序</span>
$ samtools sort -n accepted_filtered.bam accepted_sortname

<span style="color:#008000"># 注意“=”前后无空格</span>
$ htseq-count --mode=union --stranded=no --type=exon --idattr=gene_id <span style="color:#a31515">\
</span><span style="color:#a31515"></span>	          --format=bam accepted_sort.bam hg19USCS_ensembl.gtf &gt; htseqcount_accepted.hsc
</code></pre></div><p><strong>重要参数解释</strong>：</p>
<ul>
<li>
<p><code>--mode</code>：统计落在某个基因上的reads数目的模型，默认值为“union”（<a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">图示</a>）。作者认为“union”方法在绝大多数情况下都有很好的表现，建议使用。</p>
</li>
<li>
<p><code>--stranded</code>：测序方法，默认为“yes”。</p>
</li>
<li>
<p><code>--type</code>：计数单元类型，默认为<a href="http://mblab.wustl.edu/GTF22.html">GTF</a>文件的<code>exon</code>。</p>
</li>
<li>
<p><code>--idattr</code>：计数单元归类，默认为<code>gene_id</code>。比如把合并汇报多个exon对应的一个gene。</p>
</li>
<li>
<p><code>--format</code>：可以输入bam或者sam文件，bam文件需要制定此参数。</p>
</li>
</ul>
<p><strong>补充</strong>：</p>
<ul>
<li>如果输入的是bam文件，bam文件需要按照reads名称排序，操作方法为：</li>
</ul>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">$ samtools sort -n accepted_filerted.bam acceted_sortname
</code></pre></div><h3 id="12--span-stylecolorbluegenomicalignmentsspan">1.2  <!-- raw HTML omitted -->GenomicAlignments<!-- raw HTML omitted --></h3>
<p><a href="http://bioconductor.org/packages/release/bioc/html/GenomicAlignments.html">GenomicAlignments</a>是R/Bioconductor的一个包，其中<code>summarizeOverlaps</code>函数用于对alignment数据进行计数。</p>
<h3 id="13-kallisto">1.3 Kallisto</h3>
<p><strong>简介</strong>：<a href="http://pachterlab.github.io/kallisto/">kallisto</a>是由<a href="http://pachterlab.github.io/">Lior Pachter小组</a>开发一款快速测量RNA-seq数据中转录本表达丰度的软件。因为使用了<em>pseudoalignment</em>的想法，可以不用alignment，直接测量原始测序数据，因此极大提高了运算速度。因为kallisto极高的速度，可以使用<a href="http://en.wikipedia.org/wiki/Bootstrapping_(statistics)">bootstrap</a>精确估计的“不确定性（uncertainty）”，可以配合下游软件sleuth确定差异表达基因。</p>
<p><strong>平台</strong>：Linux和Mac</p>
<p><strong>快速运行</strong>：</p>
<p>首先，建立索引。</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash"><span style="color:#008000"># 查看帮助</span>
$ kallisto index 

<span style="color:#008000"># 建立索引</span>
$ kallisto index -i hg19.idx hg19UCSC_ensembl.fa
</code></pre></div><p><strong>重要参数解释</strong>：</p>
<ul>
<li>
<p><code>-i/--index</code>：kallisto输出索引文件。</p>
</li>
<li>
<ul>
<li><code>-k/--kmer-size</code>：k-mer的长度（奇数），最大值为31，默认值为31。</li>
</ul>
</li>
</ul>
<p>之后，根据原始RNA-seq数据和转录组注释，测量表达量。</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">$ kallisto quant -i hg19.idx -o kallOut -b 1000 human1.fq.gz human2.fq.gz
</code></pre></div><p><strong>重要参数解释</strong>：</p>
<ul>
<li>
<p><code>-i/--index</code>：kallisto输出索引文件。</p>
</li>
<li>
<p><code>-o/--output-dir</code>：输出目录。</p>
</li>
<li>
<p><code>-b/--bootstrap-samples</code>：bootstrap的次数，默认为0。</p>
</li>
</ul>
<h2 id="2-差异表达基因筛选">2. 差异表达基因筛选</h2>
<h3 id="21-span-stylecolorblueedgerspan">2.1 <!-- raw HTML omitted -->edgeR<!-- raw HTML omitted --></h3>
<p><strong>简介</strong>：<a href="http://www.bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>是使用RNA-seq的reads的原始计数（count）和<a href="http://en.wikipedia.org/wiki/Negative_binomial_distribution">负二项分布</a>模型，计算差异表达基因的R/Bioconductor包。edgeR能够处理RNA-seq、ChIP-seq和SAGE数据。根据edgeR的<a href="http://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">说明文件</a>数据前处理有两点要求：</p>
<ol>
<li>
<p>edgeR performs differential abundance analysis for pre-defined genomic features. Although not strictly necessary, it usually desirable that these genomic features are non-overlapping.（edgeR对预定义的基因组特征数据进行差异量分析。尽管不是严格必须，但是输入的基因特征最好能无相互重叠。）</p>
</li>
<li>
<p>The first step in an RNA-seq analysis is usually to align the raw sequence reads to a reference genome, although there are many variations on this process.（RNA-seq分析的第一步通常是把原始测序reads比对到参考基因组，当然，这一步有多种实现方法。）</p>
</li>
</ol>
<p>以上两点表明输入的数据：1. 分析某个基因组特征，比如基因、外显子、<a href="http://en.wikipedia.org/wiki/Coding_region">CDS</a>；2. 原始reads的计数，而不是一个估计量、FPKM或者RPKM，这些估计量会影响edgeR计算一些模型涉及的关键变量。对于第二点，edgeR指出也能处理类似<a href="http://deweylab.biostat.wisc.edu/rsem/">RSEM</a>输出的计数估计值，但原始计数最好。</p>
<h3 id="22-span-stylecolorbluedeseq2span">2.2 <!-- raw HTML omitted -->DESeq2<!-- raw HTML omitted --></h3>
<p><strong>简介</strong>：<a href="http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a>是R/Bioconductor的包，使用负二项分布模型和RNA-seq的原始count数值，寻找差异表达基因。<a href="http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf">DESeq2说明文件</a>指出：</p>
<ul>
<li>The count values must be raw counts of sequencing reads. This is important for DESeq2 ’s statistical model to hold, as only the actual counts allow assessing the measurement precision correctly. Hence, please do not supply other quantities, such as (rounded) normalized counts, or counts of covered base pairs – this will only lead to nonsensical results.（count数值必须为测序结果的原始count。只有使用真实的count才能正取评估模型参数，这对于DESeq2所使用的统计模型至关重要。因此，请不要使用诸如估计的（取整）count或者覆盖碱基对的count，否则只能得到没有意义的结果。）</li>
</ul>
<p><strong>平台</strong>：R跨平台使用。</p>
<p><strong>运行流程</strong>：</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-R" data-lang="R"><span style="color:#008000"># 载入包并且注册多核</span>
library(DESeq2)
library(BiocParallel)
register(MulticoreParam(4))

<span style="color:#008000"># “glioPR”是一个整理好的数据框，行表示count数，列标识样本（1~10列）和注释（11~14列）。</span>
<span style="color:#008000"># “targets”是一个数据框，表示样本的分类，注意使用因子型（factor）标记不同组</span>
&gt; targets
        Treatment  Sample Patient
human1    primary  human1       1
human2  recurrent  human2       1
human3    primary  human3       2
human4  recurrent  human4       2
human5    primary  human5       3
human6  recurrent  human6       3
human7  recurrent  human7       4
human8  recurrent  human8       5
human9    primary  human9       6
human10   primary human10       7

<span style="color:#008000"># 创建DESeqDataSet对象，使用Treatment作为比较</span>
glioPR &lt;- DESeqDataSetFromMatrix(countData = htCountSelect[, 1:10], colData = targets, design = ~ Treatment)
mcols(glioPR) &lt;- htCountSelect[, 11:14]

<span style="color:#008000"># 创建DESeqDataSet对象，使用Treatment和Pateint两个因子作为比较</span>
glioPR &lt;- DESeqDataSetFromMatrix(countData = htCountSelect[, 1:10], colData = targets, design = ~ Patient + Treatment)
mcols(glioPR) &lt;- htCountSelect[, 11:14]

<span style="color:#008000"># DEGs</span>
glioPR &lt;- DESeq(glioPR)
res &lt;- results(glioPR)
<span style="color:#008000">## 根据FDR值排序</span>
res[order(res$padj), ]
summary(res)

<span style="color:#008000"># 两种标准化count值方法， rlog（regularized log）和VST（variance stabilizing transformations）</span>
rld &lt;- rlog(glioPR)
vsd &lt;- varianceStabilizingTransformation(glioPR)
rlogMat &lt;- assay(rld)
vstMat &lt;- assay(vsd)
</code></pre></div><h3 id="参考资料">参考资料</h3>
<ul>
<li>Biconductor DESeq2 workflows: <a href="http://www.bioconductor.org/help/workflows/rnaseqGene/">RNA-Seq workflow at the gene level</a></li>
</ul>
<h3 id="更新记录">更新记录</h3>
<p>2015年6月5日</p>

		</div>

		<div class="post-tags">
			
				
					<nav class="nav tags">
							<ul class="flat">
								
								<li><a href="/tags/bioinfo">bioinfo</a></li>
								
							</ul>
					</nav>
				
			
		</div></div>
	<div class="footer wrapper">
	<nav class="nav">
		<div> © Copyright notice |  <a href="https://github.com/vividvilla/ezhil">Ezhil theme</a> | Built with <a href="https://gohugo.io">Hugo</a></div>
	</nav>
</div>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-53073828-1', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>
<script>feather.replace()</script>
</body>
</html>
